### PR TITLE
868 remove temporary ids

### LINKED ISSUE #868: Remove distinction between temporary and permanent object IDs in load balancing
**What Needs to be Done?**

**Is your feature request related to a problem? Please describe.**

**Describe potential solution outcome**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by nlslatt**
I replaced all instances of a collection element's permanent or temporary ID with a struct called `ElementIDStruct` containing an ID that is unique across ranks, the home rank, and the rank where the element currently resides.  I left the unique ID part of the struct the same as the old permanent ID, which had bits encoding the rank. There shouldn't be any logic remaining that depends on those bits, so any alternative means of making the IDs unique across ranks should work fine.

Closes #868

### CODE REVIEW

**Comment by nlslatt** [comment]
@lifflander Please update `ZoltanLB` to use the new ID struct.

File: src/vt/vrt/collection/balance/lb_common.h [hunk 1]
```diff
@@ -50,16 +50,37 @@
...
+  NodeType home_node = uninitialized_destination;
+  NodeType curr_node = uninitialized_destination;
+
+  bool operator==(const ElementIDStruct& rhs) const {
+    return id == rhs.id and home_node == rhs.home_node;
+  }
+
+  bool operator<(const ElementIDStruct& rhs) const {
+    return id < rhs.id;
+  }
```
Comment (PhilMiller) [inline src/vt/vrt/collection/balance/lb_common.h:73]: I believe for ordered collections like `std::map` and `std::set`, the definition of comparison operators needs to be consistent - i.e. I think this should be

`return id < rhs.id || (id == rhs.id && home_node < rhs.home_node);`

Comment (PhilMiller) [reply src/vt/vrt/collection/balance/lb_common.h:73]: Given the stated invariant that `id` will continue to be unique across processes, the two definitions are consistent, and just the comparison of `home_node` in `operator==` is redundant.

**Comment by PhilMiller** [comment]
I'll be very happy to see this merged. Everything looks pretty good.

File: src/vt/vrt/collection/balance/lb_common.h (hunk 1 above)
Comment (nlslatt) [reply src/vt/vrt/collection/balance/lb_common.h:74]: Redundancy removed.

**Review by PhilMiller — APPROVED** [review]

**Review by JacobDomagala — APPROVED** [review]

**Review by JacobDomagala — APPROVED** [review]
Apart from merge conflicts blocking the workflow runs, it looks good.

**Pull request merged**